Option to use HSE/LSI

seemefirst opened the pull request:
Added option to use external high speed crystal. To use the HSE/LSI a preprocessor must be defined in stm32l4xx_hal_conf.h as USE_HSE_LSI (1).

seemefirst committed:
Update system_stm32.c

Added option use HSE and LSI.  To use the HSE/LSI a USE_HSE_LSI (1) must be defined in stm32l4xx_hal_conf.h.

seemefirst committed:
Update system_stm32.c

chrismas9 commented:
This is a useful improvement as it allows for accurate timers better than 0.25%. However it is reasonable to want HSE and LSE for accurate timers and RTC. I think selection of HSx and LSx should be separate, like F4. For example PYB has HSE and LSE, but on PYBLITE the 32 kHz crystal is optional . For consistency I suggest you use the macro name `MICROPY_HW_CLK_USE_HSE`  and the existing `MICROPY_HW_RTC_USE_LSE`.

The PLL values should be defined in `mpconfigboard.h`, not in `system_stm32.c`, as some L4 run at 120 MHz, some applications run at less than 80 MHz to save power and some boards use a different crystal frequency. All the PLLSAI1 settings should use macros in `mpconfigboard.h` as the USB clock divider needs to be set to suit the crystal frequency.

When you make changes to a PR before it is reviewed you can squash the commits and force push.

seemefirst committed:
Changes made for STML4 clock usage as follows:

Added option to use HSE or MSI clock
Adjusted option to use HSE or HSI clock for RTC for STML4

Uses definition in mpconfigboard.h

clock use:

MICROPY_HW_RTC_USE_LSE  1 or 0
MICROPY_HW_CLK_USE_HSE  1 or 0

PLLSAI1 N,P,Q,R settings

define MICROPY_HW_CLK_PLLSAIN (12)
define MICROPY_HW_CLK_PLLSAIP (RCC_PLLP_DIV7)
define MICROPY_HW_CLK_PLLSAIQ (RCC_PLLQ_DIV2)
define MICROPY_HW_CLK_PLLSAIR (RCC_PLLR_DIV2)

chrismas9 commented:
@seemefirst Looks good. Thanks for making the changes.

dpgeorge commented:
Thanks for the contribution, merged in 266146ad643b408a307676e320ffb22e0fe1e0ad, with follow-up in 26e90a051415629796efbec59f1d653b46e43aea
